### PR TITLE
Esad/better inclusion proof

### LINKED ISSUE #752: Bitcoin DaVerifier use guest performant merkle verification
@ozankaymak made optimized merkle verification for bitcoin. let's use them.

### LINKED ISSUE #970: Update inclusion proof
Currnetly inclusion proof is made up of wtxids, txids and coinbase tx
We can optimize this proof by replacing txids with merkle proof of coinbase to merkle root

### PR DESCRIPTION
**Pull request by eyusufatik**
# Description

## Linked Issues
closes #970 
closes #752

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/bitcoin-da/src/helpers/merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+#[derive(Debug, Clone)]
+pub struct BitcoinMerkleTree {
+    depth: u32,
+    nodes: Vec<Vec<[u8; 32]>>,
+}
+
+impl BitcoinMerkleTree {
+    pub fn new(transactions: Vec<[u8; 32]>) -> Self {
+        // assert!(depth > 0, "Depth must be greater than 0");
```
Comment (kpp) [inline crates/bitcoin-da/src/helpers/merkle_tree.rs:11]: debug_assert!

File: crates/bitcoin-da/src/service.rs [hunk 2]
```diff
@@ -554,30 +560,44 @@ impl DaService for BitcoinService {
...
-        let coinbase_tx_hash = block.txdata[0].compute_txid().to_raw_hash().to_byte_array();
-        txids.push(coinbase_tx_hash);
-        if coinbase_tx_hash.starts_with(self.reveal_tx_id_prefix.as_slice()) {
+
+        // coinbase starts with 0, so we skip it unless the prefix is all 0's
+        if self.reveal_wtxid_prefix.iter().all(|&x| x == 0) {
             completeness_proof.push(block.txdata[0].clone());
         }
 
         block.txdata[1..].iter().for_each(|tx| {
```
Comment (kpp) [inline crates/bitcoin-da/src/service.rs:571]: start from [0..] to include coinbase tx also if prefix is 00...

File: crates/bitcoin-da/src/verifier.rs [hunk 3]
```diff
@@ -199,58 +193,52 @@ impl DaVerifier for BitcoinVerifier {
...
-                        .wtxids
-                        .iter()
-                        .map(|wtxid| Txid::from_byte_array(*wtxid));
+                    // let wtxids = inclusion_proof
+                    //     .wtxids
+                    //     .iter()
+                    //     .map(|wtxid| Wtxid::from_byte_array(*wtxid));
 
-                    let merkle_root = merkle_tree::calculate_root(wtxids).unwrap();
+                    // let merkle_root = merkle_tree::calculate_root(wtxids).unwrap();
```
Comment (kpp) [inline crates/bitcoin-da/src/verifier.rs:203]: Remove comment

File: crates/bitcoin-da/src/verifier.rs [hunk 4]
```diff
@@ -199,58 +193,52 @@ impl DaVerifier for BitcoinVerifier {
...
+            inclusion_proof.coinbase_merkle_proof,
+        );
 
         // Inclusion proof is all the txs in the block.
-        let tx_hashes = inclusion_proof.txids.into_iter().map(Txid::from_byte_array);
-
-        let Some(root_from_inclusion) = merkle_tree::calculate_root(tx_hashes) else {
-            return Err(ValidationError::FailedToCalculateMerkleRoot);
-        };
+        // let tx_hashes = inclusion_proof.txids.into_iter().map(Txid::from_byte_array);
```
Comment (kpp) [inline crates/bitcoin-da/src/verifier.rs:239]: Remove

File: crates/bitcoin-da/src/service.rs (hunk 2 above)
Comment (eyusufatik) [reply crates/bitcoin-da/src/service.rs:571]: resolving bc compute_wtxid would calc coinbases wtxid instead defaulting to 0.

so this should stay as is

File: crates/bitcoin-da/src/helpers/merkle_tree.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    //         } else {
+    //             preimage[..32].copy_from_slice(&merkle_proof[level as usize]);
+    //             preimage[32..].copy_from_slice(&combined_hash);
+    //             combined_hash = calculate_double_sha256(&preimage);
+    //         }
+    //         level += 1;
+    //         index = index / 2;
+    //     }
+    //     assert_eq!(combined_hash, self.root());
+    // }
```
Comment (rakanalh) [inline crates/bitcoin-da/src/helpers/merkle_tree.rs:127]: Cleanup

File: crates/bitcoin-da/src/service.rs [hunk 6]
```diff
@@ -427,16 +428,21 @@ impl From<TxidWrapper> for [u8; 32] {
...
+            if i == 0 {
+                // Replace the first hash with zeroes.
+                Wtxid::all_zeros().to_raw_hash().to_byte_array()
+            } else {
+                t.compute_wtxid().to_raw_hash().to_byte_array()
+            }
+        })
+        .collect();
+    BitcoinMerkleTree::new(hashes).root()
+    // merkle_tree::calculate_root(hashes).map(|h| h.into())
```
Comment (rakanalh) [inline crates/bitcoin-da/src/service.rs:445]: Same

File: crates/bitcoin-da/src/verifier.rs [hunk 7]
```diff
@@ -199,58 +193,48 @@ impl DaVerifier for BitcoinVerifier {
...
                 }
                 Some(mut commitment_idx) => {
-                    let wtxids = inclusion_proof
-                        .wtxids
-                        .iter()
-                        .map(|wtxid| Txid::from_byte_array(*wtxid));
+                    // let wtxids = inclusion_proof
+                    //     .wtxids
+                    //     .iter()
+                    //     .map(|wtxid| Wtxid::from_byte_array(*wtxid));
```
Comment (rakanalh) [inline crates/bitcoin-da/src/verifier.rs:201]: same

File: crates/bitcoin-da/src/helpers/merkle_tree.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+        // Construct the tree
+        let mut curr_level_offset: usize = 1;
+        let mut prev_level_size = transactions.len();
+        let mut prev_level_index_offset = 0;
+        let mut preimage: [u8; 64] = [0; 64];
+        while prev_level_size > 1 {
+            // println!("curr_level_offset: {}", curr_level_offset);
+            // println!("prev_level_size: {}", prev_level_size);
+            // println!("prev_level_index_offset: {}", prev_level_index_offset);
```
Comment (rakanalh) [inline crates/bitcoin-da/src/helpers/merkle_tree.rs:42]: same

**Review by rakanalh — APPROVED** [review]
LGTM

File: crates/bitcoin-da/src/service.rs [hunk 9]
```diff
@@ -554,30 +560,44 @@ impl DaService for BitcoinService {
...
+            txid_merkle_tree.root(),
+            block.header.merkle_root(),
+            "Merkle root mismatch"
+        );
+
+        let coinbase_proof = txid_merkle_tree.get_idx_path(0);
+
         (
-            InclusionMultiProof::new(txids, wtxids, block.txdata[0].clone()),
+            // todo fill merkle proof
```
Comment (ercecan) [inline crates/bitcoin-da/src/service.rs:599]: remove comment

**Review by ercecan — APPROVED** [review]

File: crates/bitcoin-da/src/helpers/merkle_tree.rs [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    nodes: Vec<Vec<[u8; 32]>>,
+}
+
+impl BitcoinMerkleTree {
+    pub fn new(transactions: Vec<[u8; 32]>) -> Self {
+        if transactions.len() == 1 {
+            // root is the coinbase txid
+            return BitcoinMerkleTree {
+                depth: 1,
+                nodes: vec![vec![transactions[0]]],
```
Comment (yaziciahmet) [inline crates/bitcoin-da/src/helpers/merkle_tree.rs:18]:
```suggestion
                nodes: vec![transactions],
```

File: crates/bitcoin-da/src/helpers/merkle_tree.rs [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+        // Populate leaf nodes
+        tree.nodes.push(vec![]);
+        for tx in transactions.iter() {
+            tree.nodes[0].push(*tx);
+        }
+
+        // Construct the tree
+        let mut curr_level_offset: usize = 1;
+        let mut prev_level_size = transactions.len();
```
Comment (yaziciahmet) [inline crates/bitcoin-da/src/helpers/merkle_tree.rs:36]:
```suggestion
        let mut tree = BitcoinMerkleTree {
            depth: depth,
            nodes: vec![transactions],
        };

        // Construct the tree
        let mut curr_level_offset: usize = 1;
        let mut prev_level_size = tree.nodes[0].len();
```

**Pull request merged**